# Exploring Ebay Car Sales Data

### A guided project using pandas.
---------------


### I. Reading the dataset and making preliminary analysis

The first thing we'll do is load the dataframe and quickly see what type of datas we have in hand, how many null values within the columns, relevance of the datas and what type of clean-up or changes that we need to make. 

In [1]:
import pandas as pd
import numpy as np 
autos = pd.read_csv("autos.csv", encoding='Windows-1252')

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

Since most of the data and a part of the column are written in Deutsch, we will later translate both the column names and the values within them into English. The values for price and mileage will also be cleaned and converted into integer.

There are a couple of items that need further analysis but some rows consist of prices that doesn't reflect a price of an automobile. 
The value of zero powerPS, PS being short for pferdestärke or horsepower, is also odd for a car as this might indicate the item listed might be parts of the car

As shown below, the null values are quite plenty but none of the columns exceeds 20%.

In [3]:
print(autos.info(),'\n')
print('Number of null values within column:')
autos.isnull().sum()
 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled               0
name                      0
seller                    0
offerType                 0
price                     0
abtest                    0
vehicleType            5095
yearOfRegistration        0
gearbox                2680
powerPS                   0
model                  2758
odometer                  0
monthOfRegistration       0
fuelType               4482
brand                     0
notRepairedDamage      9829
dateCreated               0
nrOfPictures              0
postalCode                0
lastSeen                  0
dtype: int64

### II. Initial cleanup

In the steps below, we changed the column names from camelcase to snake case as well as translating powerPS to horsepower. These changes are done to make the columns easier to understand when performing the analysis later on.

To start, we can create a list containing the column names and indices that might come in handy later on.

In [4]:
indices = []

for c in autos.columns:
    if c in autos:
        location = autos.columns.get_loc(c)
        indices.append([c,location])       
indices      

[['dateCrawled', 0],
 ['name', 1],
 ['seller', 2],
 ['offerType', 3],
 ['price', 4],
 ['abtest', 5],
 ['vehicleType', 6],
 ['yearOfRegistration', 7],
 ['gearbox', 8],
 ['powerPS', 9],
 ['model', 10],
 ['odometer', 11],
 ['monthOfRegistration', 12],
 ['fuelType', 13],
 ['brand', 14],
 ['notRepairedDamage', 15],
 ['dateCreated', 16],
 ['nrOfPictures', 17],
 ['postalCode', 18],
 ['lastSeen', 19]]

In [5]:
autos.rename(columns={
'dateCrawled':'crawled_date', 
'offerType':'offer_type', 
'vehicleType':'vechicle_type', 
'yearOfRegistration':'registration_year', 
'powerPS':'horsepower',
'monthOfRegistration':'registration_month', 
'fuelType':'fuel_type', 
'notRepairedDamage':'unrepaired_damage',
'dateCreated':'ad_created', 
'postalCode':'postal_code', 
'lastSeen':'last_seen',
'nrOfPictures':'number_of_pictures'
}, 
inplace = True)

autos.head()

crawled_date                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vechicle_type  registration_year  \
0  privat    Angebot  $5,000  control           bus               2004   
1  privat    Angebot  $8,500  control     limousine               1997   
2  privat    Angebot  $8,990     test     limousine               2009   
3  privat    Angebot  $4,350  control    kleinwagen               2007   
4  privat    Angebot  $1,350     test         kombi               2003   

     gearbox  horsepower   model   odometer  registration_month fuel_type  \
0    manuell         158  andere  150,000km                   3       lpg   
1  automatik         286     7er  150,000km                   6    benzin   
2    manuell         102    golf   70,000km                   7    benzin   
3  automatik          71  fortwo   70,000km                   6    benzin   
4    manuell           0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  number_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00                   0   
1         bmw              nein  2016-04-04 00:00:00                   0   
2  volkswagen              nein  2016-03-26 00:00:00                   0   
3       smart              nein  2016-03-12 00:00:00                   0   
4        ford              nein  2016-04-01 00:00:00                   0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

After making changes to the columns, we will also translate the values that are currently in German into English using a dictionary that will be mapped into the columns which values we want to translate.

In [6]:
en_dict = {
'bus':'bus',
'limousine':'limousine',
'kleinwagen':'small car',
'kombi':'combi',
'coupe':'coupe' ,
'suv':'suv',
'cabrio':'convertible',
'andere':'other',
'nein':'no', 
'ja':'yes',
'manuell':'manual',
'automatik':'automatic',
'lpg':'lpg',
'benzin':'gasoline',
'diesel':'diesel',
'cng':'cng',
'hybrid':'hybrid',
'elektro':'electric',     
}

autos['vechicle_type'] = autos['vechicle_type'].map(en_dict)
autos['unrepaired_damage'] = autos['unrepaired_damage'].map(en_dict)
autos['gearbox'] = autos['gearbox'].map(en_dict)
autos['fuel_type'] = autos['fuel_type'].map(en_dict)


In [7]:
autos['fuel_type']

0             lpg
1        gasoline
2        gasoline
3        gasoline
4        gasoline
5          diesel
6        gasoline
7          diesel
8             NaN
9        gasoline
10       gasoline
11       gasoline
12       gasoline
13       gasoline
14       gasoline
15       gasoline
16       gasoline
17       gasoline
18       gasoline
19            lpg
20         diesel
21       gasoline
22       gasoline
23         diesel
24       gasoline
25       gasoline
26       gasoline
27            NaN
28         diesel
29         diesel
           ...   
49970      diesel
49971    gasoline
49972      diesel
49973    gasoline
49974    gasoline
49975      hybrid
49976    gasoline
49977      diesel
49978    gasoline
49979      diesel
49980    gasoline
49981    gasoline
49982    gasoline
49983    gasoline
49984         NaN
49985    gasoline
49986      diesel
49987      diesel
49988    gasoline
49989    gasoline
49990    gasoline
49991    gasoline
49992         lpg
49993    gasoline
49994    g

### III. Basic Data Exploration?

Next stop, more data cleaning. 

In [8]:
autos.describe(include='all')

crawled_date         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-12 16:06:22  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vechicle_type  registration_year gearbox    horsepower  model  \
count          44905       50000.000000   47320  50000.000000  47242   
unique             8                NaN       2           NaN    245   
top        limousine                NaN  manual           NaN   golf   
freq           12859                NaN   36993           NaN   4024   
mean             NaN        2005.073280     NaN    116.355920    NaN   
std              NaN         105.712813     NaN    209.216627    NaN   
min              NaN        1000.000000     NaN      0.000000    NaN   
25%              NaN        1999.000000     NaN     70.000000    NaN   
50%              NaN        2003.000000     NaN    105.000000    NaN   
75%              NaN        2008.000000     NaN    150.000000    NaN   
max              NaN        9999.000000     NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN  gasoline  volkswagen                no   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  number_of_pictures   postal_code  \
count                 50000             50000.0  50000.000000   
unique                   76                 NaN           NaN   
top     2016-04-03 00:00:00                 NaN           NaN   
freq                   1946                 NaN           NaN   
mean                    NaN                 0.0  50813.627300   
std                     NaN                 0.0  25779.747957   
min                     NaN                 0.0   1067.000000   
25%                     NaN                 0.0  30451.000000   
50%                     NaN                 0.0  49577.000000   
75%                     NaN                 0.0  71540.000000   
max                     NaN                 0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

Looking at the column one by one gives us several items to highlight:
1. Price will need to be cleaned and converted to integer
2. Odometer will need to be converted to integer as well
3. horsepower has a min. of 0 (either the value is entered wrongly or this is not a car) and a max of 17700 (Danish fastest drag race car has 15,000 HP and it looks like a jet).  
4. Registration year has a bizzare min. year of 1000 and max. year of 9999



On a side note:
- Date information doesn't seem to bring any significant value to the analysis. Ad creation date might show us the high and lows of time when people are deciding to sell their cars, but we'll give this more thought later.
- Registration month is actually redundant with registration year. Not sure if it will have any added value

#### Fixing price and odometer values

In [9]:
clean_price = []
clean_odometer = []

for row in autos['price']:
    row = row.replace("$","")
    row = row.replace(",","")
    row = int(row)
    clean_price.append(row)
    
for row in autos['odometer']:
    row = row.replace("km","")
    row = row.replace(",","")
    row = int(row)
    clean_odometer.append(row)
    
autos['price'] = clean_price
autos['odometer'] = clean_odometer
autos.rename(columns={'odometer':'odometer_km'},inplace=True)

To check if it's changed to what we wanted...

In [10]:
print(autos['price'].head())
print('\n')
print(autos['odometer_km'].head())

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64


0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64


Done.

#### What about the car with 17700 horsepower?
Let's see which car is it.

In [11]:
max_hp = autos[autos["horsepower"] == 17700]
max_hp

crawled_date                      name  seller offer_type  \
36421  2016-03-20 16:51:48  Verkaufe_meinen_bmw_525d  privat    Angebot   

       price abtest vechicle_type  registration_year    gearbox  horsepower  \
36421   6000   test         combi               2004  automatic       17700   

      model  odometer_km  registration_month fuel_type brand  \
36421   5er       150000                   7    diesel   bmw   

      unrepaired_damage           ad_created  number_of_pictures  postal_code  \
36421                no  2016-03-20 00:00:00                   0        26316   

                 last_seen  
36421  2016-04-06 22:45:56

So this car in row 36421 is a BMW 525d (as in diesel engine). Looking at a [newer post of the same car model](https://m.ebay-kleinanzeigen.de/s-anzeige/bmw-verkaufe-meinen-bmw-525d-3-0-l-6-zylinder-aut-/1335283216-216-6396), we can see that the PS or HP is actually 204 and not a jawdropping 17700. 

Now, is there any other abnormally high HP value listed? Or abnormally low HP?
Let's try to look for how many rows are there with horsepower > 100 and < 500.

In [12]:
print('\n','Number of cars with horsepower between 100 to 500 =', autos['horsepower'][(autos['horsepower'] > 100) & (autos['horsepower'] < 500)].count(),'cars.')


 Number of cars with horsepower between 100 to 500 = 28871 cars.


With that approach, we ended up with with 28871 rows. If we want to take in only rows that has good horsepower data it means we will exclude  close to 43% of the data. However it doesn't mean that the remaining 57% is accurate. Given this consideration, unless we are going  to study the relation of other variables with horsepower we might as well drop the horsepower column from the analysis.

#### IV. Looking for the outliers

Before we can analyze the data, we need to clean the dataset further. We'll look into the following columns more thoroughly:
1. Price
2. Odometer
3. Registration year

What we want to see from each column is:
- Number of unique value
- Min/max/median
- Frequency value

To do this, we can set up a function that allows us to look into different columns quickly:


In [13]:
def analyze(col):
    a = print('Number of unique value in price is:', col.unique().shape,'\n')
    b = col.describe()
    c = col.value_counts().sort_index()
    d = a,b,c
    return d

prc = autos['price']
odo = autos['odometer_km']
reg = autos['registration_year']

In [14]:
analyze(prc)

Number of unique value in price is: (2357,) 



(None, count    5.000000e+04
 mean     9.840044e+03
 std      4.811044e+05
 min      0.000000e+00
 25%      1.100000e+03
 50%      2.950000e+03
 75%      7.200000e+03
 max      1.000000e+08
 Name: price, dtype: float64, 0           1421
 1            156
 2              3
 3              1
 5              2
 8              1
 9              1
 10             7
 11             2
 12             3
 13             2
 14             1
 15             2
 17             3
 18             1
 20             4
 25             5
 29             1
 30             7
 35             1
 40             6
 45             4
 47             1
 49             4
 50            49
 55             2
 59             1
 60             9
 65             5
 66             1
             ... 
 151990         1
 155000         1
 163500         1
 163991         1
 169000         1
 169999         1
 175000         1
 180000         1
 190000         1
 194000         1
 197000         1
 198000         1
 220000

As shown in the above, the price distribution is very wide with the cheapest price being 1 and the most expensive being 99,999,999 which is a range that doesn't make sense. 

Let's see how many rows remain if we set a price range between 500 and 12,000. 


In [15]:
prc_outlier = (prc >500)&(prc<12000)
prc[prc_outlier].describe()

count    37842.000000
mean      3805.425850
std       2944.463204
min        501.000000
25%       1400.000000
50%       2850.000000
75%       5550.000000
max      11999.000000
Name: price, dtype: float64

The statistics makes way more sense now compared to before we take out the outlier.

Let's do the same for odometer:

In [16]:
analyze(odo)

Number of unique value in price is: (13,) 



(None, count     50000.000000
 mean     125732.700000
 std       40042.211706
 min        5000.000000
 25%      125000.000000
 50%      150000.000000
 75%      150000.000000
 max      150000.000000
 Name: odometer_km, dtype: float64, 5000        967
 10000       264
 20000       784
 30000       789
 40000       819
 50000      1027
 60000      1164
 70000      1230
 80000      1436
 90000      1757
 100000     2169
 125000     5170
 150000    32424
 Name: odometer_km, dtype: int64)

Odometer values didn't really spell out weird number and it should be fine the way it is.

The one that needs to be cleaned would be the registration year:

In [17]:
analyze(reg)

Number of unique value in price is: (97,) 



(None, count    50000.000000
 mean      2005.073280
 std        105.712813
 min       1000.000000
 25%       1999.000000
 50%       2003.000000
 75%       2008.000000
 max       9999.000000
 Name: registration_year, dtype: float64, 1000       1
 1001       1
 1111       1
 1500       1
 1800       2
 1910       9
 1927       1
 1929       1
 1931       1
 1934       2
 1937       4
 1938       1
 1939       1
 1941       2
 1943       1
 1948       1
 1950       3
 1951       2
 1952       1
 1953       1
 1954       2
 1955       2
 1956       5
 1957       2
 1958       4
 1959       7
 1960      34
 1961       6
 1962       4
 1963       9
         ... 
 2001    2703
 2002    2533
 2003    2727
 2004    2737
 2005    3015
 2006    2708
 2007    2304
 2008    2231
 2009    2098
 2010    1597
 2011    1634
 2012    1323
 2013     806
 2014     666
 2015     399
 2016    1316
 2017    1453
 2018     492
 2019       3
 2800       1
 4100       1
 4500       1
 4800       1
 5000       4

As there are values that dates back to year 1927, even year 1001, we will only include car with registration year between 1950 to 2016.
Note that this means we will exclude the ones from year 2017, 2018 and 2019 as this data was crawled in 2016. For year 2800 and beyond, we will just ignore them.

In [18]:
reg_outlier = (reg >= 1950)&(reg <= 2016)
reg[reg_outlier].describe()

count    48004.000000
mean      2002.841513
std          7.105768
min       1950.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

Since we already have a good idea on how to identify the outliers, we'll go on ahead and remove them. We will also drop all rows that contain null value while we're at it.

In [19]:
cl_autos = autos[reg_outlier]
cl_autos = cl_autos[prc_outlier]
cl_autos = cl_autos.dropna()

/dataquest/system/env/python3/lib/python3.4/site-packages/ipykernel/__main__.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  from ipykernel import kernelapp as app


In [20]:
cl_autos

crawled_date                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
31     2016-03-14 16:53:09                          Noch_gut_erhaltenen_C_320   
33     2016-03-15 12:25:26  Audi_A3_2.0_TDI_Sportback_DPF_Kupplung_defekt_...   
35     2016-03-29 21:52:56                                        Ford_Mondeo   
36     2016-03-16 23:46:31          Ford_Mondeo_2.0_Turnier_TDCi_DPF_Titanium   
38     2016-03-21 15:51:10                        Volkswagen_Golf_1.4_Special   
39     2016-03-11 13:58:38                 Mercedes_C_180_Kombi__zu_verkaufen   
40     2016-03-07 14:50:03  VW_Golf__4_Cabrio_2.0_GTI_16V___Leder_MFA_Alus...   
43     2016-03-08 16:51:22           Audi_A1_1.2_TFSI_Attraction_mit_Garantie   
44     2016-03-12 10:38:06                                       Zu_verkaufen   
45     2016-03-12 17:39:55                                 Toyota_Starlet_1.3   
47     2016-03-28 19:51:51                                Mercedes_Benz_A_150   
48     2016-03-20 12:41:50                           Mercedes_C180_Sport_1_8l   
50     2016-03-09 16:48:05  •??_????????_??•Top!!!_Peugeot_308_sw_Sport_•?...   
51     2016-03-22 12:57:38                            Mercedes_Benz_S_320_CDI   
53     2016-04-02 17:45:01      Dacia_Sandero_Stepway_II_MPI_85__top_gepflegt   
54     2016-03-15 19:51:34                              Ford_Focus_1.8_Futura   
...                    ...                                                ...   
49940  2016-03-11 14:52:16                                      Omas_Liebling   
49942  2016-03-09 19:43:33             Audi_A3_2.0_TDI_Sportback_DPF_Ambition   
49944  2016-03-15 17:48:16  SAAB_9_3__2_0t_SE__Automatik__Tempomat__Regens...   
49945  2016-04-04 12:54:29                  omega_2_2_facelift_mit_neuem_tuev   
49946  2016-03-05 18:52:14                                 Opel_Kadett_SR_1.3   
49949  2016-03-26 20:48:26  VW_Golf_3__1.4__60_PS__Ratte__Rost__09/16_TÜV_...   
49951  2016-03-31 17:54:48  Opel_Corsa_D1.0_weiss_/Scheckheftgepflegt/_Unf...   
49952  2016-03-21 21:45:33  VW_POLO__9N3_Silber_1_2_Ltr._5_tuerer__96_tsd....   
49956  2016-03-28 22:39:12        Toyota_Yaris_1.3_VVT_i_Executive_MODEL_2007   
49959  2016-03-28 21:53:21           Mercedes_Benz_C_180_T_Kompressor_Classic   
49962  2016-03-14 17:57:15  Mitsubishi_Space_Star_1__3_L__Bj_2004_Standhei...   
49963  2016-03-26 19:52:59            Mercedes_Benz_B_200_CDI_Special_Edition   
49965  2016-03-11 10:53:51          Opel_Astra_1.6_Lenkradheizung~Sitzheizung   
49966  2016-04-02 19:49:19  Citroën_C1_1.0_**Euro4**TÜV_OKT_2017**Scheiten...   
49968  2016-04-01 17:49:15                  Mercedes_Benz_190_D_2.5_Automatik   
49969  2016-03-17 18:49:02             Nissan_X_Trail_2.2_dCi_4x4_Sport_m.AHZ   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_

We have 27134 remaining rows of clean data that we can now analyze. 
On to the next!


### V. Wait, now we're working on the dates?

We've managed to cover the registration year and registration month. We'll now work on the date_crawled, ad_created and last_seen.

In [21]:
date_var = cl_autos[['crawled_date','ad_created','last_seen']]
date_var.head()

crawled_date           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [22]:
autos['crawled_date'].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05 14:06:30    0.00002
2016-03-05 14:06:40    0.00002
2016-03-05 14:07:04    0.00002
2016-03-05 14:07:08    0.00002
2016-03-05 14:07:21    0.00002
2016-03-05 14:07:26    0.00002
2016-03-05 14:07:40    0.00002
2016-03-05 14:07:45    0.00002
2016-03-05 14:08:00    0.00004
2016-03-05 14:08:05    0.00004
2016-03-05 14:08:27    0.00002
2016-03-05 14:08:42    0.00002
2016-03-05 14:09:02    0.00004
2016-03-05 14:09:05    0.00002
2016-03-05 14:09:20    0.00002
2016-03-05 14:09:22    0.00002
2016-03-05 14:09:38    0.00002
2016-03-05 14:09:46    0.00002
2016-03-05 14:09:56    0.00002
2016-03-05 14:09:57    0.00002
2016-03-05 14:09:58    0.00004
2016-03-05 14:10:18    0.00002
2016-03-05 14:10:20    0.00002
2016-03-05 14:10:46    0.00002
2016-03-05 14:11:03    0.00002
2016-03-05 14:11:05    0.00002
2016-03-05 14:11:14    0.00002
2016-03-05 14:11:15    0.00002
2016-03-05 14:11:25    0.00002
2016-03-05 14:11:40    0.00002
                        ...   
2016-04-07 10:36:19    0.00002
2016-04-

In [23]:
autos['ad_created'].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11 00:00:00    0.00002
2015-08-10 00:00:00    0.00002
2015-09-09 00:00:00    0.00002
2015-11-10 00:00:00    0.00002
2015-12-05 00:00:00    0.00002
2015-12-30 00:00:00    0.00002
2016-01-03 00:00:00    0.00002
2016-01-07 00:00:00    0.00002
2016-01-10 00:00:00    0.00004
2016-01-13 00:00:00    0.00002
2016-01-14 00:00:00    0.00002
2016-01-16 00:00:00    0.00002
2016-01-22 00:00:00    0.00002
2016-01-27 00:00:00    0.00006
2016-01-29 00:00:00    0.00002
2016-02-01 00:00:00    0.00002
2016-02-02 00:00:00    0.00004
2016-02-05 00:00:00    0.00004
2016-02-07 00:00:00    0.00002
2016-02-08 00:00:00    0.00002
2016-02-09 00:00:00    0.00004
2016-02-11 00:00:00    0.00002
2016-02-12 00:00:00    0.00006
2016-02-14 00:00:00    0.00004
2016-02-16 00:00:00    0.00002
2016-02-17 00:00:00    0.00002
2016-02-18 00:00:00    0.00004
2016-02-19 00:00:00    0.00006
2016-02-20 00:00:00    0.00004
2016-02-21 00:00:00    0.00006
                        ...   
2016-03-09 00:00:00    0.03324
2016-03-

In [24]:
autos['last_seen'].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05 14:45:46    0.00002
2016-03-05 14:46:02    0.00002
2016-03-05 14:49:34    0.00002
2016-03-05 15:16:11    0.00002
2016-03-05 15:16:47    0.00002
2016-03-05 15:28:10    0.00002
2016-03-05 15:41:30    0.00002
2016-03-05 15:45:43    0.00002
2016-03-05 15:47:38    0.00002
2016-03-05 15:47:44    0.00002
2016-03-05 16:45:57    0.00002
2016-03-05 16:47:28    0.00002
2016-03-05 17:15:45    0.00002
2016-03-05 17:16:12    0.00002
2016-03-05 17:16:14    0.00002
2016-03-05 17:16:23    0.00002
2016-03-05 17:17:02    0.00002
2016-03-05 17:39:19    0.00002
2016-03-05 17:40:14    0.00002
2016-03-05 17:44:50    0.00002
2016-03-05 17:44:54    0.00002
2016-03-05 17:46:01    0.00002
2016-03-05 18:17:58    0.00002
2016-03-05 18:47:14    0.00002
2016-03-05 18:50:38    0.00002
2016-03-05 19:15:08    0.00002
2016-03-05 19:15:20    0.00002
2016-03-05 19:15:42    0.00002
2016-03-05 19:16:36    0.00002
2016-03-05 19:17:17    0.00002
                        ...   
2016-04-07 14:58:09    0.00004
2016-04-

After reviewing the distribution of crawled_date, ad_created and last_seen, we can conclude that there are no value in analyzing them. The only date variable that holds important value is registration_year which we have covered in the previous section.

In [25]:
cl_autos['registration_year'].value_counts(normalize=True)

2004    0.075662
2005    0.074740
2003    0.074740
2006    0.074114
2001    0.068438
1999    0.067369
2002    0.066890
2000    0.065821
2007    0.058414
2008    0.052222
1998    0.050785
2009    0.047247
1997    0.037186
2010    0.028673
1996    0.025872
2011    0.023550
1995    0.018980
2012    0.017837
1994    0.011867
1992    0.007481
2013    0.007445
1993    0.007297
1991    0.006744
1990    0.005270
2014    0.004459
1989    0.003464
1988    0.002617
1986    0.001585
1987    0.001511
1984    0.001179
2015    0.001179
1985    0.001032
1983    0.000995
1982    0.000774
1980    0.000774
1978    0.000627
1979    0.000479
1981    0.000479
1972    0.000369
1975    0.000369
1973    0.000332
1976    0.000332
1968    0.000332
1977    0.000332
1965    0.000295
1969    0.000258
1970    0.000258
1967    0.000184
2016    0.000184
1964    0.000184
1974    0.000147
1961    0.000111
1971    0.000111
1963    0.000074
1956    0.000074
1960    0.000074
1966    0.000074
1962    0.000037
1950    0.0000

Top 10 distribution for the car registration year is as follow:
1. 2004
2. 2005
3. 2003
4. 2006
5. 2001
6. 1999
7. 2002
8. 2000
9. 2007
10. 2008

Which is 2/3 of the whole car being on sale.

In [26]:
year = [0.075662,
0.074740,
0.074740,
0.074114,
0.068438,
0.067369,
0.066890,
0.065821,
0.058414,
0.052222]
sum(year)


0.67841

### VI. Aggregation analysis

The analysis that we'll be working on will first take into account all brands to find the most expensive brand. From there, we will focus on analyzing that particular brand



In [27]:
cl_autos['brand'].value_counts(normalize=True)

volkswagen       0.209479
opel             0.112516
bmw              0.106324
mercedes_benz    0.092909
audi             0.075993
ford             0.070023
renault          0.049458
peugeot          0.034201
fiat             0.026019
seat             0.020454
skoda            0.019496
citroen          0.016769
mazda            0.016437
toyota           0.016142
nissan           0.016105
smart            0.015958
hyundai          0.011498
volvo            0.010467
mini             0.008550
mitsubishi       0.008440
honda            0.008440
kia              0.007924
alfa_romeo       0.007592
suzuki           0.007408
chevrolet        0.006155
chrysler         0.004165
dacia            0.003649
daihatsu         0.002506
jeep             0.002174
subaru           0.002027
saab             0.001880
daewoo           0.001511
land_rover       0.001364
rover            0.001290
jaguar           0.001179
porsche          0.001032
lancia           0.000958
trabant          0.000848
lada        

In [28]:
brand_dict = {}
cl_brand = ['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'citroen', 'mazda', 'toyota',
       'nissan', 'smart', 'hyundai', 'volvo', 'mini', 'mitsubishi', 'honda',
       'kia', 'alfa_romeo', 'suzuki', 'chevrolet', 'chrysler', 'dacia',
       'daihatsu', 'jeep', 'subaru', 'saab', 'daewoo', 'land_rover', 'rover',
       'jaguar', 'porsche', 'lancia', 'trabant', 'lada']

for e in cl_brand:
    mean_price = cl_autos['price'].loc[cl_autos['brand'] == e].mean()
    brand_dict[e]=mean_price
    
brand_mean = pd.DataFrame.from_dict(brand_dict, orient='index')
brand_mean.columns = ['mean price']
brand_mean.head(10).sort_values('mean price',ascending=False)

mean price
audi        5183.306499
land_rover  5145.324324
seat        3953.893694
honda       3929.423581
subaru      3455.490909
chrysler    3372.513274
fiat        3343.869688
opel        3123.183099
saab        2863.235294
renault     2709.274963

It's not surprising to see that porsche is the most expensive brand in the listing as most people will associate porsche with famous and expensive sport car. It has been well established since 1931 in Germany and has distinguished itself for so many years with high quality parts and manufacturing. 

In [29]:
cl_porsche = cl_autos[cl_autos['brand'] == 'porsche']
cl_porsche['model'].replace('andere','other', inplace=True)

porsche_mdl = ['cayenne', 'boxster', 'other', '911']
porsche_dict = {}
for e in porsche_mdl:
    mean_price = cl_porsche['price'][cl_porsche['model'] == e].mean()
    porsche_dict[e] = mean_price

porsche_model = pd.DataFrame.from_dict(porsche_dict, orient='index')
porsche_model.columns=['mean price']
porsche_model.sort_values('mean price', ascending=False)
    

/dataquest/system/env/python3/lib/python3.4/site-packages/pandas/core/generic.py:4619: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._update_inplace(new_data)


mean price
boxster  10047.375000
cayenne   9249.727273
911       7925.000000
other     5309.714286

We can see in the dataframe above, the mean prices for all 4 porsche models available in the listing.

Price for lower mileage cars for all brands and types is 1.65 times more expensive compared to that of a higher mileage car. But what is interesting, this doesn't apply to porsche.

In [30]:
high_mileage = []
for row in cl_autos['odometer_km']:
    if row > 75000:
        high_mileage.append('high mileage')
    else:
        high_mileage.append('low mileage')
cl_autos['high_mileage'] = high_mileage

Lower mileage most of the time correlates with a newer car or more favorable condition compared to the higher mileage. Therefore it is expected to see the numbers as below:

In [31]:
mileage_dict = {}
cl_mlg = ['high mileage', 'low mileage']

for e in cl_mlg:
    mean_price = cl_autos['price'].loc[(cl_autos['high_mileage'] == e)].mean()
    mileage_dict[e]=mean_price
    
mlg_prc = pd.DataFrame.from_dict(mileage_dict, orient='index')
mlg_prc.columns = ['general mean price']
mlg_prc

general mean price
low mileage          6386.511531
high mileage         3865.819592

In [32]:
mileage_dict = {}
cl_mlg = ['high mileage', 'low mileage']

for e in cl_mlg:
    mean_price = cl_autos['price'].loc[(cl_autos['brand'] == 'porsche') & (cl_autos['high_mileage'] == e)].mean()
    mileage_dict[e]=mean_price
    
mlg_prc = pd.DataFrame.from_dict(mileage_dict, orient='index')
mlg_prc.columns = ['porsche mean price']
mlg_prc.sort_values('porsche mean price', ascending=False)

porsche mean price
high mileage         8663.760000
low mileage          6183.333333

Porsche mean price doesn't seem to be affected by this. It seems that even higher mileage porsche can still have high selling price.

Of course this would need to be reviewed further if a certain model is responsible for the higher price despite the higher mileage.

In [33]:
odometer_dict = {}
cl_brand = ['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'citroen', 'mazda', 'toyota',
       'nissan', 'smart', 'hyundai', 'volvo', 'mini', 'mitsubishi', 'honda',
       'kia', 'alfa_romeo', 'suzuki', 'chevrolet', 'chrysler', 'dacia',
       'daihatsu', 'jeep', 'subaru', 'saab', 'daewoo', 'land_rover', 'rover',
       'jaguar', 'porsche', 'lancia', 'trabant', 'lada']

for e in cl_brand:
    mean_odo = cl_autos['odometer_km'].loc[cl_autos['brand'] == e].mean()
    odometer_dict[e]=mean_odo
    
odo_mean = pd.DataFrame.from_dict(odometer_dict, orient='index')
odo_mean.columns = ['mean mileage']
pd.concat([odo_mean, brand_mean],axis=1).sort_values('mean price', ascending=False).head(6)

mean mileage   mean price
porsche  125535.714286  8398.000000
mini     107715.517241  7193.646552
jeep     145084.745763  6256.101695
dacia     87070.707071  5449.050505
skoda    115604.914934  5232.827977
audi     143021.338506  5183.306499

For a better comparison, we can see that with the top 6 most expensive priced brands, there is really no distinct relation between mean mileage and mean price. 